File: backend/routes_python/generate.py
from fastapi import APIRouter, HTTPException, UploadFile, File, Form, BackgroundTasks
from pydantic import BaseModel
from typing import Optional, List
import uuid
import asyncio
import json
from datetime import datetime
import aiofiles
import os
from pathlib import Path

# Import the real AI service
from services.aiService import aiService

router = APIRouter(prefix="/generate", tags=["generation"])

# Pydantic models
class GenerateRequest(BaseModel):
    prompt: str
    mode: str = "text-to-image"
    sessionId: Optional[str] = None

class GenerateResponse(BaseModel):
    success: bool
    message: str
    generationId: str
    status: str
    estimatedTime: str

class GenerationStatus(BaseModel):
    success: bool
    generation: dict


@router.post("/", response_model=GenerateResponse)
async def generate_image(request: GenerateRequest, background_tasks: BackgroundTasks):
    """Generate image from text prompt"""
    try:
        # Validate input
        if not request.prompt or len(request.prompt.strip()) < 3:
            raise HTTPException(status_code=400, detail="Prompt must be at least 3 characters")
        
        if len(request.prompt) > 500:
            raise HTTPException(status_code=400, detail="Prompt must be less than 500 characters")
        
        # Generate session ID if not provided
        session_id = request.sessionId or str(uuid.uuid4())
        
        # Create generation record
        generation_id = str(uuid.uuid4())
        generation_data = {
            "_id": generation_id,
            "prompt": request.prompt,
            "mode": request.mode,
            "sessionId": session_id,
            "status": "processing",
            "createdAt": datetime.utcnow(),
            "outputImages": [],
            "processingTime": None
        }
        
        # In a real app, save to database here
        # await db.generations.insert_one(generation_data)
        
        # Start background processing
        background_tasks.add_task(process_generation, generation_id, request.prompt, request.mode)
        
        return GenerateResponse(
            success=True,
            message="Generation started",
            generationId=generation_id,
            status="processing",
            estimatedTime="0.8-2 seconds"
        )
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

async def process_generation(generation_id: str, prompt: str, mode: str):
    """Background task to process image generation"""
    try:
        # Use real AI processing
        result = await aiService.generateImage(prompt, mode)
        
        # In a real app, update database here
        # Update generation status to completed with results
        print(f"Generation {generation_id} completed: {result['images'][0]}")
        
    except Exception as e:
        print(f"Generation {generation_id} failed: {str(e)}")
        # In a real app, update database status to failed

@router.get("/{generation_id}", response_model=GenerationStatus)
async def get_generation_status(generation_id: str):
    """Get generation status and result"""
    try:
        # In a real app, fetch from database
        # For demo, return mock completed status
        mock_generation = {
            "_id": generation_id,
            "status": "completed",
            "prompt": "A beautiful landscape",
            "outputImages": ["https://images.unsplash.com/photo-1506905925346-21bda4d32df4?w=800&q=80"],
            "processingTime": 1200,
            "createdAt": datetime.utcnow().isoformat()
        }
        
        return GenerationStatus(success=True, generation=mock_generation)
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.post("/upload")
async def upload_reference_image(
    image: UploadFile = File(...),
    sessionId: str = Form(...)
):
    """Upload reference image for image-to-image generation"""
    try:
        # Validate file
        if not image.content_type.startswith('image/'):
            raise HTTPException(status_code=400, detail="Only image files are allowed")
        
        if image.size > 10 * 1024 * 1024:  # 10MB
            raise HTTPException(status_code=400, detail="File too large. Maximum size is 10MB")
        
        # Generate unique filename
        file_extension = Path(image.filename).suffix
        filename = f"{uuid.uuid4()}{file_extension}"
        
        # Ensure uploads directory exists
        uploads_dir = Path(__file__).parent.parent / "uploads"
        uploads_dir.mkdir(exist_ok=True)
        
        # Save file
        file_path = uploads_dir / filename
        async with aiofiles.open(file_path, 'wb') as f:
            content = await image.read()
            await f.write(content)
        
        return {
            "success": True,
            "message": "Image uploaded successfully",
            "file": {
                "id": filename.replace(file_extension, ""),
                "fileName": filename,
                "originalName": image.filename,
                "size": len(content),
                "url": f"/api/files/{filename}",
                "uploadedAt": datetime.utcnow().isoformat()
            }
        }
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Upload failed: {str(e)}")

@router.get("/history/{session_id}")
async def get_generation_history(session_id: str, limit: int = 20, skip: int = 0):
    """Get generation history for a session"""
    try:
        # In a real app, fetch from database
        # For demo, return empty history
        return {
            "success": True,
            "generations": [],
            "pagination": {
                "total": 0,
                "limit": limit,
                "skip": skip,
                "hasMore": False
            }
        }
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

File: backend/routes/generate.js
const express = require('express');
const multer = require('multer');
const { v4: uuidv4 } = require('uuid');
const ImageGeneration = require('../models/ImageGeneration');
const Gallery = require('../models/Gallery');
const aiService = require('../services/aiService');
const fileService = require('../services/fileService');

const router = express.Router();

// Configure multer for file uploads
const upload = multer({
  storage: multer.memoryStorage(),
  limits: {
    fileSize: 10 * 1024 * 1024, // 10MB
    files: 9 // Maximum 9 files for batch processing
  },
  fileFilter: (req, file, cb) => {
    if (file.mimetype.startsWith('image/')) {
      cb(null, true);
    } else {
      cb(new Error('Only image files are allowed'));
    }
  }
});

/**
 * Generate image from text prompt
 * POST /api/generate
 */
router.post('/', async (req, res) => {
  try {
    const { prompt, mode = 'text-to-image', sessionId } = req.body;

    // Validate input
    const errors = ImageGeneration.validate({ prompt, sessionId, mode });
    if (errors.length > 0) {
      return res.status(400).json({ 
        success: false, 
        error: 'Validation failed', 
        details: errors 
      });
    }

    // Additional AI service validation
    const aiErrors = aiService.validateGenerationParams(prompt, mode);
    if (aiErrors.length > 0) {
      return res.status(400).json({ 
        success: false, 
        error: 'Invalid parameters', 
        details: aiErrors 
      });
    }

    // Create generation record
    const generation = new ImageGeneration({
      prompt,
      mode,
      sessionId: sessionId || uuidv4(),
      status: 'processing'
    });

    // Save to database
    const result = await req.db.collection('generations').insertOne(generation.toJSON());
    generation._id = result.insertedId;

    // Start AI generation (async)
    setImmediate(async () => {
      try {
        const aiResult = await aiService.generateImage(prompt, mode);
        
        if (aiResult.success) {
          generation.updateStatus('completed', aiResult.processingTime);
          generation.outputImages = aiResult.images;
          
          // Update in database
          await req.db.collection('generations').updateOne(
            { _id: generation._id },
            { $set: generation.toJSON() }
          );

          console.log('Generation completed:', generation._id.toString());
        } else {
          generation.updateStatus('failed');
          await req.db.collection('generations').updateOne(
            { _id: generation._id },
            { $set: generation.toJSON() }
          );
          console.error('Generation failed:', aiResult.error);
        }
      } catch (error) {
        console.error('Generation processing error:', error);
        generation.updateStatus('failed');
        await req.db.collection('generations').updateOne(
          { _id: generation._id },
          { $set: generation.toJSON() }
        );
      }
    });

    // Return immediate response with generation ID
    res.status(202).json({
      success: true,
      message: 'Generation started',
      generationId: generation._id.toString(),
      status: 'processing',
      estimatedTime: '0.8-2 seconds'
    });

  } catch (error) {
    console.error('Generate route error:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Internal server error',
      message: error.message 
    });
  }
});

/**
 * Get generation status and result
 * GET /api/generate/:id
 */
router.get('/:id', async (req, res) => {
  try {
    const { id } = req.params;

    // Find generation in database
    const generation = await req.db.collection('generations').findOne({ 
      _id: req.db.ObjectId(id) 
    });

    if (!generation) {
      return res.status(404).json({ 
        success: false, 
        error: 'Generation not found' 
      });
    }

    res.json({
      success: true,
      generation: generation
    });

  } catch (error) {
    console.error('Get generation error:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Internal server error',
      message: error.message 
    });
  }
});

/**
 * Upload reference image for image-to-image generation
 * POST /api/generate/upload
 */
router.post('/upload', upload.single('image'), async (req, res) => {
  try {
    const { sessionId } = req.body;
    
    if (!req.file) {
      return res.status(400).json({ 
        success: false, 
        error: 'No image file provided' 
      });
    }

    if (!sessionId) {
      return res.status(400).json({ 
        success: false, 
        error: 'Session ID is required' 
      });
    }

    // Save uploaded file
    const fileInfo = await fileService.saveUploadedFile(req.file, sessionId);

    res.json({
      success: true,
      message: 'Image uploaded successfully',
      file: {
        id: fileInfo.id,
        fileName: fileInfo.fileName,
        originalName: fileInfo.originalName,
        size: fileInfo.size,
        url: fileInfo.url,
        uploadedAt: fileInfo.uploadedAt
      }
    });

  } catch (error) {
    console.error('Upload error:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Upload failed',
      message: error.message 
    });
  }
});

/**
 * Get generation history for a session
 * GET /api/generate/history/:sessionId
 */
router.get('/history/:sessionId', async (req, res) => {
  try {
    const { sessionId } = req.params;
    const { limit = 20, skip = 0 } = req.query;

    const generations = await req.db.collection('generations')
      .find({ sessionId })
      .sort({ createdAt: -1 })
      .limit(parseInt(limit))
      .skip(parseInt(skip))
      .toArray();

    const total = await req.db.collection('generations').countDocuments({ sessionId });

    res.json({
      success: true,
      generations,
      pagination: {
        total,
        limit: parseInt(limit),
        skip: parseInt(skip),
        hasMore: (parseInt(skip) + parseInt(limit)) < total
      }
    });

  } catch (error) {
    console.error('History route error:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Internal server error',
      message: error.message 
    });
  }
});

/**
 * Add generation to public gallery
 * POST /api/generate/:id/gallery
 */
router.post('/:id/gallery', async (req, res) => {
  try {
    const { id } = req.params;
    const { title, description } = req.body;

    // Validate input
    if (!title || title.trim().length < 3) {
      return res.status(400).json({ 
        success: false, 
        error: 'Title must be at least 3 characters' 
      });
    }

    // Find the generation
    const generationData = await req.db.collection('generations').findOne({ 
      _id: req.db.ObjectId(id) 
    });

    if (!generationData) {
      return res.status(404).json({ 
        success: false, 
        error: 'Generation not found' 
      });
    }

    const generation = new ImageGeneration(generationData);

    if (!generation.isCompleted()) {
      return res.status(400).json({ 
        success: false, 
        error: 'Generation is not completed yet' 
      });
    }

    // Create gallery item
    const galleryItem = Gallery.createFromGeneration(
      generation, 
      title.trim(), 
      description ? description.trim() : `Generated with: "${generation.prompt}"`
    );

    // Save to gallery collection
    const result = await req.db.collection('gallery').insertOne(galleryItem.toJSON());
    galleryItem._id = result.insertedId;

    res.status(201).json({
      success: true,
      message: 'Added to gallery successfully',
      galleryItem: galleryItem.toJSON()
    });

  } catch (error) {
    console.error('Gallery add error:', error);
    res.status(500).json({ 
      success: false, 
      error: 'Internal server error',
      message: error.message 
    });
  }
});

module.exports = router;